File: 0145/iterative.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <stack>
#include <vector>

#include "leetcode/tree.hpp"

using namespace std;
using namespace LeetCode;

class Solution {
 public:
  vector<int> postorderTraversal(TreeNode* root) {
    vector<int> ret;
    if (!root) return ret;
    stack<TreeNode*> node_stack;
    node_stack.push(root);
    TreeNode* node;
    while (!node_stack.empty()) {
      node = node_stack.top();
      node_stack.pop();
      ret.push_back(node->val);
      if (node->left) node_stack.push(node->left);
      if (node->right) node_stack.push(node->right);
    }
    reverse(ret.begin(), ret.end());
    return ret;
  }
};

int main() {
  Solution sol;
  string str;
  TreeNode* root;
  vector<int> got, expected;

  root = CreateTreeNode("1 # 2 # # 3 ;");
  expected = {3,2,1};
  got = sol.postorderTraversal(root);
  assert(got == expected);
  DeleteTreeNode(root);

  //    1
  //  2   3
  // 4 5 6 7
  root = CreateTreeNode("1 2 3 4 5 6 7 ;");
  expected = {4,5,2,6,7,3,1};
  got = sol.postorderTraversal(root);
  assert(got == expected);
  DeleteTreeNode(root);

  got = sol.postorderTraversal(nullptr);
  assert(got.size() == 0);
}

File: 0107/iterative.py
from queue import Queue
from typing import *


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        q = Queue()
        ret = []
        if root is None:
            return ret
        q.put((root, 0))
        while not q.empty():
            node, level = q.get()
            if len(ret) <= level:
                ret.insert(0, [node.val])
            else:
                ret[-1-level].append(node.val)
            if node.left is not None:
                q.put((node.left, level+1))
            if node.right is not None:
                q.put((node.right, level + 1))
        return ret


if __name__ == '__main__':
    sol = Solution()

    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)
    assert sol.levelOrderBottom(root) == [[15,7],[9,20],[3]]

    assert sol.levelOrderBottom(None) == []
